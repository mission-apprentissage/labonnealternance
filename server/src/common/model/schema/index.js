import { apiCallSchema } from "./ApiCall.js"
import { applicationSchema } from "./applications.js"
import { appointmentSchema } from "./appointment.js"
import { credentialSchema } from "./credentials.js"
import { diplomesMetiersSchema } from "./diplomesmetiers.js"
import { domainesMetiersSchema } from "./domainesmetiers.js"
import { etablissementSchema } from "./etablissement.js"
import { formulaireSchema } from "./formulaire.js"
import { bonneBoiteSchema } from "./lbb/bonneBoite.js"
import { emailBlacklist } from "./lbb/emailBlacklist.js"
import { geoLocationSchema } from "./lbb/geoLocations.js"
import { opco } from "./lbb/opco.js"
import { mnaFormationSchema } from "./mnaFormation/mnaFormation.js"
import { offreSchema } from "./offre.js"
import { optoutSchema } from "./optout.js"
import { parcoursupEtablissementStatSchema } from "./parcoursupEtablissementStat.js"
import { userRecruteurSchema } from "./recruteur/users.js"
import { referentielOpcoSchema } from "./referentielOpco.js"
import { userSchema } from "./user.js"
import { widgetParameterSchema } from "./widgetParameter.js"

export default {
  domainesMetiersSchema,
  diplomesMetiersSchema,
  applicationSchema,
  apiCallSchema,
  mnaFormationSchema,
  bonneBoiteSchema,
  geoLocationSchema,
  emailBlacklist,
  opco,
  userSchema,
  appointmentSchema,
  widgetParameterSchema,
  etablissementSchema,
  parcoursupEtablissementStatSchema,
  formulaireSchema,
  offreSchema,
  referentielOpcoSchema,
  optoutSchema,
  userRecruteurSchema,
  credentialSchema,
}
