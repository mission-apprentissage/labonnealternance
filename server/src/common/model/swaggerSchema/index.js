import requireAll from "require-all"
import __dirname from "../../dirname.js"

let models = {}

function filterFile(filename) {
  if (filename.endsWith(".js") && filename !== "index.js") {
    return filename
  }

  return false
}

const modelsList = requireAll({
  dirname: __dirname(import.meta.url),
  filter: (filename) => filterFile(filename),
})

Object.keys(modelsList).forEach((filename) => {
  const model = modelsList[filename]
  models = { ...models, ...model }
})

export default models
