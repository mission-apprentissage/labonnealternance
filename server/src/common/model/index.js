import mongoosePaginate from "mongoose-paginate"
import { getElasticInstance, mongoosastic } from "../esClient/index.js"
import schema from "../model/schema/index.js"
import { mongooseInstance } from "../mongodb.js"

const createModel = (modelName, descriptor, options = {}) => {
  const schema = new mongooseInstance.Schema(descriptor)

  if (options.esIndexName) {
    schema.plugin(mongoosastic, { esClient: getElasticInstance(), index: options.esIndexName })
  }

  if (options.esIndexName || options.paginate) {
    schema.plugin(mongoosePaginate)
  }

  if (options.createMongoDBIndexes) {
    options.createMongoDBIndexes(schema)
  }

  return mongooseInstance.model(modelName, schema)
}

const DomainesMetiers = createModel("domainesmetiers", schema.domainesMetiersSchema, {
  esIndexName: "domainesmetiers",
})
const FormationCatalogue = createModel("formationcatalogues", schema.mnaFormationSchema, {
  esIndexName: "formationcatalogues",
})
const DiplomesMetiers = createModel("diplomesmetiers", schema.diplomesMetiersSchema, {
  esIndexName: "diplomesmetiers",
})
const ApiCalls = createModel("apicalls", schema.apiCallSchema)
const Application = createModel("applications", schema.applicationSchema, {
  paginate: true,
})
const SourceFormations = createModel("sourceformations", schema.sourceFormationsSchema)
const GeoLocation = createModel("geolocation", schema.geoLocationSchema)
const EmailBlacklist = createModel("emailblacklist", schema.emailBlacklist)
const Opco = createModel("opco", schema.opco)
const BonnesBoites = createModel("bonnesboites", schema.bonneBoiteSchema, {
  esIndexName: "bonnesboites",
})
const User = createModel("user", schema.userSchema, { paginate: true })
const Appointment = createModel("appointment", schema.appointmentSchema, { paginate: true })
const WidgetParameter = createModel("widgetParameter", schema.widgetParameterSchema, { paginate: true })
const Etablissement = createModel("etablissement", schema.etablissementSchema, { paginate: true })
const ParcoursupEtablissementStat = createModel("parcoursupEtablissementStat", schema.parcoursupEtablissementStatSchema, { paginate: true })
const Formulaire = createModel("formulaire", schema.formulaireSchema, { esIndexName: "formulaires", paginate: true })
const Credential = createModel("credentials", schema.credentialSchema)
const Offre = createModel("offre", schema.offresSchema)
const Optout = createModel("optout", schema.optoutSchema, { paginate: true })
const ReferentielOpco = createModel("referentielOpco", schema.referentielOpcoSchema)
const UserRecruteur = createModel("userRecruteur", schema.userRecruteurSchema, { paginate: true })
const AppointmentDetailed = createModel("appointmentDetailed", schema.appointmentSchema)

export {
  DomainesMetiers,
  FormationCatalogue,
  DiplomesMetiers,
  ApiCalls,
  Application,
  SourceFormations,
  GeoLocation,
  EmailBlacklist,
  Opco,
  BonnesBoites,
  User,
  Appointment,
  WidgetParameter,
  Etablissement,
  ParcoursupEtablissementStat,
  Formulaire,
  Credential,
  Offre,
  Optout,
  ReferentielOpco,
  UserRecruteur,
  AppointmentDetailed,
}
