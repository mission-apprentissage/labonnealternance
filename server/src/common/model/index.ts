import { captureException } from "@sentry/node"

import { db, mongooseInstance } from "@/common/mongodb"

import { logger } from "../logger"

import AnonymizedUser from "./schema/anonymizedUsers/anonymizedUsers.schema"
import ApiCalls from "./schema/apiCall/apiCall.schema"
import AnonymizedApplication from "./schema/application/anonymizedApplications.schema"
import Application from "./schema/application/applications.schema"
import AppointmentDetailed from "./schema/appointmentDetailed/appointmentDetailed.schema"
import Appointment from "./schema/appointments/appointment.schema"
import Credential from "./schema/credentials/credential.schema"
import CustomEmailETFA from "./schema/customEmailETFA/customEmailETFA.schema"
import DiplomesMetiers from "./schema/diplomesmetiers/diplomesmetiers.schema"
import DomainesMetiers from "./schema/domainesmetiers/domainesmetiers.schema"
import EligibleTrainingsForAppointment from "./schema/eligibleTrainingsForAppointment/eligibleTrainingsForAppointment.schema"
import eligibleTrainingsForAppointmentHistory from "./schema/eligibleTrainingsForAppointmentsHistory/eligibleTrainingsForAppointmentHistory.schema"
import EmailBlacklist from "./schema/emailBlacklist/emailBlacklist.schema"
import Etablissement from "./schema/etablissements/etablissement.schema"
import FicheMetierRomeV4 from "./schema/ficheRomeV4/ficheRomeV4"
import FormationCatalogue from "./schema/formationCatalogue/formationCatalogue.schema"
import GeoLocation from "./schema/geolocation/geolocation.schema"
import InternalJobs from "./schema/internalJobs/internalJobs.schema"
import Job from "./schema/jobs/jobs.schema"
import LbaCompany from "./schema/lbaCompany/lbaCompany.schema"
import LbaCompanyLegacy from "./schema/lbaCompanylegacy/lbaCompanyLegacy.schema"
import { Cfa } from "./schema/multiCompte/cfa.schema"
import { Entreprise } from "./schema/multiCompte/entreprise.schema"
import { RoleManagement } from "./schema/multiCompte/roleManagement.schema"
import { UserWithAccount } from "./schema/multiCompte/userWithAccount.schema"
import Opco from "./schema/opco/opco.schema"
import Optout from "./schema/optout/optout.schema"
import Recruiter from "./schema/recruiter/recruiter.schema"
import RecruteurLbaUpdateEvent from "./schema/recruteurLbaUpdateEvent/recruteurLbaUpdateEvent.schema"
import ReferentielOnisep from "./schema/referentielOnisep/referentielOnisep.schema"
import ReferentielOpco from "./schema/referentielOpco/referentielOpco.schema"
import ReferentielRome from "./schema/referentielRome/referentielRome"
import Session from "./schema/session/session.schema"
import SiretDiffusibleStatus from "./schema/siretDiffusibleStatusSchema/siretDiffusibleStatusSchema.schema"
import UnsubscribedLbaCompany from "./schema/unsubscribedLbaCompany/unsubscribedLbaCompany.schema"
import UnsubscribeOF from "./schema/unsubscribedOF/unsubscribeOF.schema"
import User from "./schema/user/user.schema"

const createSpecialIndexes = async () => {
  await db.collection("bonnesboites").createIndex({ geopoint: "2dsphere" })
  await db.collection("formationcatalogues").createIndex({ lieu_formation_geopoint: "2dsphere" })
  await db.collection("recruiters").createIndex({ geopoint: "2dsphere" })
}

export async function createMongoDBIndexes() {
  const results = await Promise.allSettled(
    mongooseInstance.modelNames().map(async (name) => {
      const model = mongooseInstance.model(name)
      return model.createIndexes({ background: true }).catch(async (e) => {
        if (e.codeName === "IndexOptionsConflict" || e.codeName === "IndexKeySpecsConflict") {
          const err = new Error(`Conflict in indexes for ${name}`, { cause: e })
          logger.error(err)
          captureException(err)
          await mongooseInstance.connection.collection(model.collection.name).dropIndexes()
          await model.createIndexes({ background: true })
        } else {
          throw e
        }
      })
    })
  )

  await createSpecialIndexes()

  const errors = results.reduce((acc, r) => {
    if (r.status === "rejected") {
      acc.push(r.reason)

      logger.error(r.reason)
      captureException(r.reason)
    }

    return acc
  }, [] as Error[])

  if (errors.length > 0) {
    throw new AggregateError(errors, `createMongoDBIndexes failed with ${errors.length} errors`)
  }
}

export {
  AnonymizedApplication,
  AnonymizedUser,
  ApiCalls,
  Application,
  Appointment,
  AppointmentDetailed,
  Cfa,
  Credential,
  CustomEmailETFA,
  DiplomesMetiers,
  DomainesMetiers,
  EligibleTrainingsForAppointment,
  EmailBlacklist,
  Entreprise,
  Etablissement,
  FicheMetierRomeV4,
  FormationCatalogue,
  GeoLocation,
  InternalJobs,
  Job,
  LbaCompany,
  LbaCompanyLegacy,
  RecruteurLbaUpdateEvent,
  Opco,
  Optout,
  Recruiter,
  ReferentielOnisep,
  ReferentielOpco,
  ReferentielRome,
  RoleManagement,
  Session,
  SiretDiffusibleStatus,
  UnsubscribeOF,
  UnsubscribedLbaCompany,
  User,
  UserWithAccount,
  eligibleTrainingsForAppointmentHistory,
}
