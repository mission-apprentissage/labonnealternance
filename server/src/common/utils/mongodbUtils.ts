import { captureException } from "@sentry/node"
import { isEqual } from "lodash-es"
import { Collection, CollectionInfo, MongoClient, MongoServerError } from "mongodb" // to uncomment when migrated to V7
import { IModelDescriptor } from "shared/models/common"
import { CollectionName, IDocument, modelDescriptors } from "shared/models/models"
import { zodToMongoSchema } from "zod-mongodb-schema"

import config from "../../config"
import { logger } from "../logger"

import { sleep } from "./asyncUtils"

let mongodbClient: MongoClient | null = null
let mongodbClientState: string | null = null

export const ensureInitialization = () => {
  if (!mongodbClient) {
    throw new Error("Database connection does not exist. Please call connectToMongodb before.")
  }
  return mongodbClient
}

/**
 * @param  {string} uri
 * @returns client
 */
export const connectToMongodb = async (uri: string) => {
  const client = new MongoClient(uri, {
    heartbeatFrequencyMS: 10_000,
    retryWrites: true,
    retryReads: true,
    minPoolSize: config.env === "local" ? 0 : 5,
    maxPoolSize: 1_000,
  })

  client.on("connectionPoolReady", () => {
    logger.info("MongoDB reconnected")
    mongodbClient = client
  })

  client.on("connectionPoolClosed", () => {
    logger.warn("MongoDB closed")
    mongodbClient = null
  })

  await client.connect()
  // @ts-expect-error
  mongodbClientState = client.topology.s.state
  mongodbClient = client
  logger.info("Connected to MongoDB")

  return client
}

export const getMongodbClient = () => mongodbClient
export const getMongodbClientState = () => mongodbClientState

export const closeMongodbConnection = async () => {
  logger.warn("Closing MongoDB")
  // Let 100ms for possible callback cleanup to register tasks in mongodb queue
  await sleep(200)
  return mongodbClient?.close()
}

export const getDatabase = () => {
  return ensureInitialization().db()
}

export const getDbCollection = <K extends CollectionName>(name: K): Collection<IDocument<K>> => {
  return ensureInitialization().db().collection(name)
}

export const getCollectionList = () => {
  return ensureInitialization().db().listCollections().toArray()
}

export const getDbCollectionIndexes = async (name: CollectionName) => {
  return await ensureInitialization().db().collection(name).indexes()
}

/**
 * Création d'une collection si elle n'existe pas
 * @param {string} collectionName
 */
const createCollectionIfDoesNotExist = async (collectionName: string) => {
  const db = getDatabase()
  const collectionsInDb = await db.listCollections().toArray()
  const collectionExistsInDb = collectionsInDb.map(({ name }) => name).includes(collectionName)

  if (!collectionExistsInDb) {
    try {
      await db.createCollection(collectionName)
    } catch (err) {
      if ((err as MongoServerError).codeName !== "NamespaceExists") {
        throw err
      }
    }
  }
}

/**
 * Vérification de l'existence d'une collection à partir de la liste des collections
 * @param {*} collectionsInDb
 * @param {*} collectionName
 * @returns
 */
export const collectionExistInDb = (collectionsInDb: CollectionInfo[], collectionName: string) => collectionsInDb.map(({ name }: { name: string }) => name).includes(collectionName)

/**
 * Config de la validation
 * @param {*} modelDescriptors
 */
export const configureDbSchemaValidation = async (modelDescriptors: IModelDescriptor[]) => {
  const db = getDatabase()
  ensureInitialization()
  await Promise.all(
    modelDescriptors.map(async ({ collectionName, zod }) => {
      await createCollectionIfDoesNotExist(collectionName)

      const convertedSchema = zodToMongoSchema(zod)

      try {
        await db.command({
          collMod: collectionName,
          validationLevel: "strict",
          validationAction: "error",
          validator: {
            $jsonSchema: {
              title: `${collectionName} validation schema`,
              ...convertedSchema,
            },
          },
        })
      } catch (error) {
        captureException(error)
        logger.error(error)
      }
    })
  )
}

/**
 * Clear de toutes les collections
 * @returns
 */
export const clearAllCollections = async () => {
  const collections = await getDatabase().collections()
  return Promise.all(collections.map((c) => c.deleteMany({})))
}

/**
 * Clear d'une collection
 * @param {string} name
 * @returns
 */
export async function clearCollection(name: string) {
  ensureInitialization()
  await getDatabase().collection(name).deleteMany({})
}

export const createIndexes = async () => {
  for (const descriptor of modelDescriptors) {
    if (!descriptor.indexes) {
      return
    }
    const indexes = await getDbCollection(descriptor.collectionName)
      .listIndexes()
      .toArray()
      .catch((err) => {
        // NamespaceNotFound
        if (err.code === 26) {
          return []
        }
        throw err
      })
    const indexesToRemove = new Set(indexes.filter((i) => i.name !== "_id_"))

    logger.info(`Create indexes for collection ${descriptor.collectionName}`)
    await Promise.all(
      descriptor.indexes.map(async ([index, options]): Promise<void> => {
        try {
          const existingIndex =
            // Use Object.entries because order matters
            indexes.find((i) => isEqual(Object.entries(i.key), Object.entries(index)) || options.name === i.name) ?? null

          if (existingIndex) {
            indexesToRemove.delete(existingIndex)
          }

          await getDbCollection(descriptor.collectionName)
            .createIndex(index, options)
            .catch(async (err) => {
              // IndexOptionsConflict & IndexKeySpecsConflict
              if (err.code === 85 || err.code === 86) {
                await getDbCollection(descriptor.collectionName).dropIndex(existingIndex.name)
                await getDbCollection(descriptor.collectionName).createIndex(index, options)
              } else {
                throw err
              }
            })
        } catch (err) {
          captureException(err)
          logger.error(`Error creating indexes for ${descriptor.collectionName}: ${err}`)
        }
      })
    )

    if (indexesToRemove.size > 0) {
      logger.warn(`Dropping extra indexes for collection ${descriptor.collectionName}`, indexesToRemove)
      await Promise.all(Array.from(indexesToRemove).map((index) => getDbCollection(descriptor.collectionName).dropIndex(index.name)))
    }
  }
}

export const dropIndexes = async () => {
  const collections = (await getCollectionList()).map((collection) => collection.name)
  for (const descriptor of modelDescriptors) {
    logger.info(`Drop indexes for collection ${descriptor.collectionName}`)
    if (collections.includes(descriptor.collectionName)) {
      await getDbCollection(descriptor.collectionName).dropIndexes()
    }
  }
}
