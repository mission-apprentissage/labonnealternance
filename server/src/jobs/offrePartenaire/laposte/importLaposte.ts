import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"
import rawLaposteModel from "shared/models/rawLaposte.model"

import config from "../../../config"
import { importFromStreamInXml } from "../importFromStreamInXml"
import { importFromUrlInXml } from "../importFromUrlInXml"
import { rawToComputedJobsPartners } from "../rawToComputedJobsPartners"

import { laposteJobToJobsPartners, ZLaposteJob } from "./laposteMapper"

const rawCollectionName = rawLaposteModel.collectionName
const offerXmlTag = "offre"

export const importLaposteRaw = async (sourceStream?: NodeJS.ReadableStream) => {
  if (sourceStream) {
    await importFromStreamInXml({
      destinationCollection: rawCollectionName,
      offerXmlTag,
      stream: sourceStream,
      partnerLabel: JOBPARTNERS_LABEL.LAPOSTE,
      conflictingOpeningTagWithoutAttributes: true,
    })
  } else {
    await importFromUrlInXml({
      destinationCollection: rawCollectionName,
      url: config.laposteUrl,
      offerXmlTag,
      partnerLabel: JOBPARTNERS_LABEL.LAPOSTE,
      conflictingOpeningTagWithoutAttributes: true,
    })
  }
}

export const importLaposteToComputed = async () => {
  await rawToComputedJobsPartners({
    collectionSource: rawCollectionName,
    partnerLabel: JOBPARTNERS_LABEL.LAPOSTE,
    zodInput: ZLaposteJob,
    mapper: laposteJobToJobsPartners,
    documentJobRoot: offerXmlTag,
  })
}
