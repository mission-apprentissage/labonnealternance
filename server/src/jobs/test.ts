import { ObjectId } from "bson"
import { IApplicant, IApplication, ZApplicant } from "shared"

import { getDbCollection } from "@/common/utils/mongodbUtils"

import { asyncForEach } from "../common/utils/asyncUtils"

type ApplicationAggregate = Pick<IApplication, "applicant_email" | "applicant_first_name" | "applicant_last_name" | "applicant_phone" | "_id">

export const migrationTest = async () => {
  // console.log(extensions.telephone.safeParse("0690760552"))
  // return
  const applications = (await getDbCollection("applications")
    .find({}, { projection: { applicant_first_name: 1, applicant_last_name: 1, applicant_email: 1, applicant_phone: 1, _id: 1 } })
    .toArray()) as ApplicationAggregate[]

  const stat = { error: 0, success: 0, total: applications.length }

  await asyncForEach(applications, async (application) => {
    const { applicant_first_name, applicant_email, applicant_last_name, applicant_phone, _id } = application
    const now = new Date()
    const applicant: IApplicant = {
      _id: new ObjectId(),
      firstname: applicant_first_name,
      lastname: applicant_last_name,
      phone: applicant_phone,
      email: applicant_email,
      last_connection: null,
      createdAt: now,
      updatedAt: now,
    }
    const validation = ZApplicant.safeParse(applicant)
    if (!validation.success) {
      stat.error++
      console.log({ applicant, err: validation.error.errors })
    } else {
      stat.success++
    }
  })
  console.log(stat)
}
