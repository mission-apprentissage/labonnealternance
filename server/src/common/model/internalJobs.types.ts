import { ObjectId } from "mongodb"

import { CronName } from "@/jobs/jobs"

/*
  name: Le nom de la tâche
  type: Type du job simple ou cron
  status: Statut courant du job
  sync: Si le job est synchrone
  payload: La donnée liéé à la tâche
  output: Les valeurs de retours du job
  cron_string: standard cron string exemple: '*\/2 * * * *'
  scheduled_for: Date de lancement programmée
  started_at: Date de lancement
  ended_at: Date de fin d'execution
  updated_at: Date de mise à jour en base de données
  created_at: Date d'ajout en base de données
*/

interface IInternalJobsSimple {
  _id: ObjectId
  name: string
  type: "simple"
  status: "pending" | "will_start" | "running" | "finished" | "blocked" | "errored"
  sync: boolean
  payload: Record<string, any>
  output?: Record<string, unknown> | undefined
  scheduled_for: Date
  started_at?: Date
  ended_at?: Date
  updated_at: Date
  created_at: Date
}

interface IInternalJobsCron {
  _id: ObjectId
  name: CronName
  type: "cron"
  status: "pending" | "will_start" | "running" | "finished" | "blocked" | "errored"
  sync: boolean
  cron_string: string
  scheduled_for: Date
  updated_at: Date
  created_at: Date
}

interface IInternalJobsCronTask {
  _id: ObjectId
  name: CronName
  type: "cron_task"
  status: "pending" | "will_start" | "running" | "finished" | "blocked" | "errored"
  sync: boolean
  scheduled_for: Date
  started_at?: Date
  ended_at?: Date
  updated_at: Date
  created_at: Date
}

type IInternalJobs = IInternalJobsSimple | IInternalJobsCron | IInternalJobsCronTask

export type { CronName, IInternalJobs, IInternalJobsCron, IInternalJobsCronTask, IInternalJobsSimple }
