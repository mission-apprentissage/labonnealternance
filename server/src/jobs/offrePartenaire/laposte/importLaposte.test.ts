import fs from "node:fs"

import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"
import { beforeEach, describe, expect, it, vi } from "vitest"

import { getDbCollection } from "@/common/utils/mongodbUtils"
import { useMongo } from "@tests/utils/mongo.test.utils"

import { importLaposteRaw, importLaposteToComputed } from "./importLaposte"

const now = new Date("2024-07-21T04:49:06.000+02:00")

describe("importLaposte", () => {
  useMongo()

  beforeEach(() => {
    vi.useFakeTimers()
    vi.setSystemTime(now)

    return async () => {
      vi.useRealTimers()
      await getDbCollection("computed_jobs_partners").deleteMany({})
      await getDbCollection("raw_hellowork").deleteMany({})
    }
  })

  it("should test the import of La Poste data into computed_job_partners", async () => {
    const fileStream = fs.createReadStream("server/src/jobs/offrePartenaire/laposte/importLaposte.test.input.xml")
    await importLaposteRaw(fileStream)
    expect.soft(await getDbCollection("raw_laposte").countDocuments({})).toBe(2)

    await importLaposteToComputed()
    const jobs = (
      await getDbCollection("computed_jobs_partners")
        .find({ partner_label: JOBPARTNERS_LABEL.LAPOSTE }, { projection: { _id: 0, created_at: 0 } })
        .toArray()
    ).sort((a, b) => ((a.partner_job_id ?? "") < (b.partner_job_id ?? "") ? -1 : 1))
    expect.soft(jobs.length).toBe(2)
    expect.soft(jobs).toMatchSnapshot()
  })
})
