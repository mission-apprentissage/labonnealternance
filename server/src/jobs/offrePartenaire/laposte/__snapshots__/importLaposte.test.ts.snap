// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importLaposte > should test the import of La Poste data into computed_job_partners 1`] = `
[
  {
    "apply_phone": null,
    "apply_url": "https://www.laposterecrute.fr/offre/2025-xxxxx1",
    "business_error": null,
    "contract_duration": 12,
    "contract_remote": "onsite",
    "contract_start": null,
    "contract_type": [
      "Apprentissage",
    ],
    "errors": [],
    "jobs_in_success": [],
    "offer_access_conditions": [],
    "offer_creation": 2025-05-22T00:11:18.000Z,
    "offer_description": "- Métier : Responsable commercial
- Fiche métier : https://www.laposterecrute.fr/fichemetier/responsable-commercial
- Filière : Commercial
- Expérience : 0-1 an
- Temps de travail hebdomadaire : 35

Description de la mission :

Vous souhaitez acquérir de l'expérience dans le développement commercial B to B, développer le chiffre d’affaires d’une zone de marché ? Animer des équipes autour des produits et techniques de vente ? La Poste Groupe et son CFA Formaposte Sud Est vous proposent un parcours en alternance de 12 mois, qui combine la préparation d’un Bachelor Responsable en Développement Commercial et Marketing avec IFC Valence reconnu par le Ministère du Travail de Niveau 6 (Bac+3), avec l’apprentissage du métier de Commercial B to B en plateforme de distribution courrier. Les missions du Commercial B to B: Vous commercialisez les offres courrier/colis auprès de client professionnel Vous renforcez la connaissance des clients du Groupe La Poste pour fidéliser et conquérir de nouveaux prospects. Vous pilotez les indicateurs commerciaux pour atteindre les objectifs de votre secteur Vous animez les équipes managériales et opérationnelles autour des nouveaux services pour une atteinte collective du chiffre d’affaires Vous assurez la bonne remontée d’informations entre les différentes équipes de votre secteur Vous accompagnez les équipes dans les espaces de vente pour développer et personnaliser la relation client

Profil candidat :

Afin d'être éligible pour cette formation, vous devez : Vous êtes titulaire (ou en cours d'obtention) d'un diplôme de niveau BAC+2 Etre éligible au contrat d'apprentissage (- de 30 ans /pas de limite d'âge pour les personnes possédant une RQTH) Justifier d'une première expérience professionnelle Permis B souhaité. Intégrer un centre de formation, Formaposte Sud Est et obtenir un Bachelor Responsable en Développement Commercial avec  IFC Valence. Intégrer une entreprise, La Poste et bénéficier d'une rémunération (entre 1 000 € et 1790 €* BRUT selon votre âge et votre contrat).  * Salaire donné à titre indicatif, peut varier lors de votre entrée en formation",
    "offer_desired_skills": [],
    "offer_expiration": 2025-07-22T00:11:18.000Z,
    "offer_multicast": true,
    "offer_opening_count": 1,
    "offer_origin": null,
    "offer_rome_codes": null,
    "offer_status": "Active",
    "offer_status_history": [],
    "offer_target_diploma": {
      "european": "6",
      "label": "Licence, Maîtrise, autres formations niveaux 6 (Bac+3 à Bac+4)",
    },
    "offer_title": "Alternance Commercial B to B F/H",
    "offer_to_be_acquired_knowledge": [],
    "offer_to_be_acquired_skills": [],
    "partner_job_id": "2025-XXXX1",
    "partner_label": "La Poste",
    "updated_at": 2024-07-21T02:49:06.000Z,
    "validated": false,
    "workplace_address_city": "Aubenas",
    "workplace_address_label": "Aubenas",
    "workplace_address_street_label": null,
    "workplace_address_zipcode": null,
    "workplace_brand": null,
    "workplace_description": "Service : autre compagnie

La Poste Groupe change, nos métiers évoluent. Etre toujours au plus près des Français, développer la confiance dans le numérique et être acteur de la transformation écologique, c'est aussi le sens de notre métier. Rejoindre La Poste Groupe, c'est rejoindre une entreprise responsable, riche de ses 232 000 collaborateurs ! Pour l'égalité des chances, La Poste fait vivre la diversité. Nos postes sont ouverts à toutes et à tous. Vous aussi, engagez-vous à nos côtés pour donner du sens à votre métier.",
    "workplace_geopoint": {
      "coordinates": [
        4.591775,
        43.612671,
      ],
      "type": "Point",
    },
    "workplace_idcc": null,
    "workplace_legal_name": null,
    "workplace_naf_code": null,
    "workplace_naf_label": null,
    "workplace_name": "La Poste",
    "workplace_opco": null,
    "workplace_siret": null,
    "workplace_size": null,
    "workplace_website": null,
  },
  {
    "apply_phone": null,
    "apply_url": "https://www.laposterecrute.fr/offre/2025-xxxxx9",
    "business_error": null,
    "contract_duration": 24,
    "contract_remote": null,
    "contract_start": null,
    "contract_type": [
      "Apprentissage",
    ],
    "errors": [],
    "jobs_in_success": [],
    "offer_access_conditions": [],
    "offer_creation": 2025-05-22T00:11:19.000Z,
    "offer_description": "- Métier : Responsable collecte distribution
- Fiche métier : https://www.laposterecrute.fr/fichemetier/responsable-collecte-distribution
- Filière : Distribution / Livraison
- Expérience : 0-1 an
- Temps de travail hebdomadaire : 35

Description de la mission :

Vous souhaitez acquérir de l'expérience dans le management, l’accompagnement au changement, l’amélioration de l’expérience client en lien avec les équipes de votre site ? Vous cherchez un job au cœur de l’action, dynamique, polyvalent ? La Poste Groupe et son CFA Formaposte vous proposent un parcours en alternance de 24 mois, qui combine la

Profil candidat :

Afin d'être éligible pour cette formation, vous devez :  Être éligible au contrat d'apprentissage (- de 30 ans/pas de limite d'âge pour les personnes possédant une RQTH)  Être titulaire d'un BAC+3 Être mobile au sein des lieux où les postes sont à pourvoir Expérience professionnelle de 6 mois Permis B souhaité. Intégrer un centre de formation, Formaposte Sud Est et obtenir Mastère Manager de Projets Innovants &amp; Entrepreneuriat en partenariat avec La Poste et bénéficier d'une rémunération (entre 1 467 € et 2 067 €* BRUT selon votre âge et votre contrat). * Salaire donné à titre indicatif, peut varier lors de votre entrée en formation.",
    "offer_desired_skills": [],
    "offer_expiration": 2025-07-22T00:11:19.000Z,
    "offer_multicast": true,
    "offer_opening_count": 1,
    "offer_origin": null,
    "offer_rome_codes": null,
    "offer_status": "Active",
    "offer_status_history": [],
    "offer_target_diploma": {
      "european": "7",
      "label": "Master, titre ingénieur, autres formations niveau (Bac+5)",
    },
    "offer_title": "Alternance Responsable production F/H",
    "offer_to_be_acquired_knowledge": [],
    "offer_to_be_acquired_skills": [],
    "partner_job_id": "2025-xxxx9",
    "partner_label": "La Poste",
    "updated_at": 2024-07-21T02:49:06.000Z,
    "validated": false,
    "workplace_address_city": "La Seyne-sur-Mer",
    "workplace_address_label": "La Seyne-sur-Mer",
    "workplace_address_street_label": null,
    "workplace_address_zipcode": null,
    "workplace_brand": null,
    "workplace_description": "Service : nom compagnie

La Poste Groupe change, nos métiers évoluent. Etre toujours au plus près des Français, développer la confiance dans le numérique et être acteur de la transformation écologique, c'est aussi le sens de notre métier. Rejoindre La Poste Groupe, c'est rejoindre une entreprise responsable, riche de ses 232 000 collaborateurs ! Pour l'égalité des chances, La Poste fait vivre la diversité. Nos postes sont ouverts à toutes et à tous. Vous aussi, engagez-vous à nos côtés pour donner du sens à votre métier.",
    "workplace_geopoint": {
      "coordinates": [
        4.979651,
        42.193147,
      ],
      "type": "Point",
    },
    "workplace_idcc": null,
    "workplace_legal_name": null,
    "workplace_naf_code": null,
    "workplace_naf_label": null,
    "workplace_name": "La Poste",
    "workplace_opco": null,
    "workplace_siret": null,
    "workplace_size": null,
    "workplace_website": null,
  },
]
`;
